### ISSUE TITLE
Feature request: Ability to change snapshot name format

### ISSUE THREAD

**Issue by darkpixel** [jimsalterjrs/sanoid#552]
I would love the ability to change the snapshot name format.

It's a minor nitpick, but if you put these two snapshot names in to your terminal:
```
tank/officeshare@zfs-auto-snap_04-2020-03-01-1600
tank/officeshare@syncoid_usrbgofnas01_2020-05-01:20:42:59
```
...and then try to double-click each one to select it, the snapshot created by zfs-auto-snapshot automatically selects the entire name.  The snapshot by sanoid only selects ```tank/officeshare@syncoid_usrbgofnas01_2020-05-01``` because it runs into a colon.

I would love the ability to change the snapshot name format so date and time are separated by an underscore and the the individual fields are separated by dashes.

**Comment by Yvan-Masson** [comment]
Hi,

This feature would be essential for my use case: I need to share snapshots via SMB (Samba) to Windows computers, which is currently impossible because the colon is a forbidden character. Currently, snapshot name is mangled by Samba, which makes it meaningless.

Regards,
Yvan

**Comment by phreaker0** [comment]
@Yvan-Masson you can use the samba catia module to remap the illegal chars, see: https://github.com/jimsalterjrs/sanoid/issues/345#issuecomment-459660254

**Comment by Yvan-Masson** [comment]
@phreaker0 Thanks for the tip! I have also just seen samba shadow_copy2 module which might not be trivial to setup but could do the job (https://github.com/jimsalterjrs/sanoid/issues/5#issuecomment-730660664).

**Comment by reefland** [comment]
It would be helpful to alter snapshot name to remove the seconds from the snapshot name.  I have TrueNAS pulling Sanoid snapshots and TrueNAS does not allow seconds as a variable in schema name. So I have to hard code that as ":00", but often Sanoid second is at 01 which is annoying because we don't need seconds in a snapshot name.

**Comment by jimsalterjrs** [comment]
I'm not sure what you mean by "TrueNAS pulling Sanoid snapshots." Can you elaborate?

**Comment by reefland** [comment]
Jim, be glad to explain it...

TrueNAS (previously known as FreeNAS) has the ability to PUSH out to other servers and/or PULL in from other servers;  ZFS snapshots using its built in "Replication Tasks".  Behind the scenes the SSH session happens and the normal ZFS send / receive happens.

When defining a Replication Task on TrueNAS to PULL snapshots created by sainoid on an Ubuntu 20.04.1 system, besides defining the source ZFS pool and dataset(s) to pull from I have to define a Naming Schema to parse the snapshot names to define a filter of which snapshots to import.  They look like:

`autosnap_%Y-%m-%d_%H:%M:00_hourly`
`autosnap_%Y-%m-%d_%H:%M:00_daily`
`autosnap_%Y-%m-%d_%H:%M:00_monthly`

As you will notice, I have to hard code the ":00" for seconds as TrueNAS does provide a variable placeholder to represent the seconds.  So I want to be able to configure Sanoid to not have seconds in the snapshot name.  They are useless,  Nobody is doing multiple snapshots per minute.

The problems happens when Saniod is 1 second late and the snapshot ends with ":01" TrueNAS can't see this snapshot and thus can not pull it.  I only want accuracy to the minute, not the second.

Hope that was helpful.

**Comment by darkpixel** [comment]
Have you already tried ```autosnap_%Y-%m-%d_%H:%M:%S_hourly```?
The ```%S``` is [strptime](https://linux.die.net/man/3/strptime) for seconds.

**Comment by jimsalterjrs** [comment]
According to TrueNAS docs at https://www.ixsystems.com/documentation/freenas/11.3-U3.2/tasks.html#replication-creation-wizard, TrueNAS naming schema uses FreeBSD strftime(), which links to manpage here: https://www.freebsd.org/cgi/man.cgi?query=strftime

strftime() supports %S for seconds in the same format Sanoid uses. From where I'm sitting, if TrueNAS doesn't support %S, that's a TrueNAS bug which should be reported there. (They really, REALLY ought to support wildcard use as well, but that's a slightly different kettle of fish I suppose.)

**Comment by reefland** [comment]
You're right, looking back I did a typo of "%s" not "%S".  So I can make it work, tested this morning it works.

I would not have opened a TrueNAS bug to fight for the ability to have seconds when I'm trying to get rid of them.  To me it makes no sense to have seconds in the schema name when Sanoid is doing snapshots at most every 15 minutes.

My request still stands to be able to alter the Sanoid snapshot name format to get rid of the seconds.

**Comment by reefland** [comment]
BTW - @jimsalterjrs thank you for your work on Sanoid.  Great product.

**Comment by reefland** [comment]
Last comment, I noticed this in the defaults conf:
```
# for snapshots shorter than one hour, the period duration must be defined
# in minutes. Because they are executed within a full hour, the selected
# value should divide 60 minutes without remainder so taken snapshots
# are apart in equal intervals. Values larger than 59 aren't practical
# as only one snapshot will be taken on each full hour in this case.
```
I take that to mean there isn't a way to get more than 1 snapshot per minute, so accuracy down to the seconds is irrelevant.  Not seeing the value of it being in the name.

**Comment by darkpixel** [comment]
@reefland In some cases the seconds field is annoying (i.e. when trying to integrate snapshots with Samba's VFS module), but I don't think chopping it off would be a good idea.  Being precise is good.  But I totally agree the snapshot names should be modifiable.  I can't tell you how much I despise colons in the snapshot name.  When double-clicking on them in a Linux terminal to highlight text, it stops at colons meaning I have to click and drag to select instead of double-clicking. It's a small thing, but it's annoying.

zfs-auto-snapshot does names like this: ```tank@zfs-auto-snap_hourly-2021-02-02-0217```.  Easily double-clickable to select. ;)

**Comment by reefland** [comment]
zfs auto snapshot as well as TrueNas does not include seconds in the snapshot name.  :)

**Comment by darkpixel** [comment]
True.  I suppose I could grab it from the ```creation``` property.

**Comment by reefland** [comment]
One last request to be able to change snapshot name.   All my `_monthly` snapshots failed to replicate to TrueNAS this month because the snapshot name was 1 seconds ahead of the snapshot creation time.  

If I could remove the seconds from the snapshot name I could resolve this issue.  Perhaps sanoid naming snapshots that don't match the creation time is real bug, not sure.

Here is the failure from TrueNAS trying to pull in this snapshot:
```
* Replication "tank/docker_projects -> main/docker/docker_backups" failed: zettarepl: zfs send PID is 852589
warning: cannot send 'tank/docker_projects/certbot@autosnap_2021-03-01_00:00:01_monthly': not an earlier snapshot from the same fs
Replication cannot continue because existing snapshot
autosnap_2021-03-01_00:00:01_hourly is newer than
'tank/docker_projects/certbot@autosnap_2021-03-01_00:00:01_monthly', but has an older date
in the snapshot name. To resolve the error, rename
'tank/docker_projects/certbot@autosnap_2021-03-01_00:00:01_monthly' with a date that is older than
autosnap_2021-03-01_00:00:01_hourly or delete snapshot
'tank/docker_projects/certbot@autosnap_2021-03-01_00:00:01_monthly' from both the source and destination..
```
Snapshot name is one second newer than snapshot creation time:
```
$ zfs get creation tank/docker_projects/certbot@autosnap_2021-03-01_00:00:01_monthly
NAME                                                               PROPERTY  VALUE                  SOURCE
tank/docker_projects/certbot@autosnap_2021-03-01_00:00:01_monthly  creation  Mon Mar  1  0:00 2021  
```
Rename snapshot to match:
```
sudo zfs rename tank/docker_projects/certbot@autosnap_2021-03-01_00:00:01_monthly tank/docker_projects/certbot@autosnap_2021-03-01_00:00:00_monthly

# Verified name matches creation time.
$ zfs get creation tank/docker_projects/certbot@autosnap_2021-03-01_00:00:00_monthly
NAME                                                               PROPERTY  VALUE                  SOURCE
tank/docker_projects/certbot@autosnap_2021-03-01_00:00:00_monthly  creation  Mon Mar  1  0:00 2021  -
```
The replication error clears...
```
The following alert has been cleared:
* Replication "tank/docker_projects -> main/docker/docker_backups" failed: zettarepl: zfs send PID is 852589
warning: cannot send 'tank/docker_projects/certbot@autosnap_2021-03-01_00:00:01_monthly': not an earlier snapshot from the same fs
Replication cannot continue because existing snapshot....
```

Now I have to find all snapshots matching "@autosnap_2021-03-01_00:00:01_monthly" and rename them to be 1 second sooner.

**Comment by jimsalterjrs** [comment]
I strongly suggest you file a bug with TrueNAS re: its odd refusal to replicate snapshots based on their names. That's not a ZFS thing... that's a TrueNAS thing.

**Comment by reefland** [comment]
I've left a message in the form, we'll see if it gets any traction.  I agree it's not a ZFS thing.   

But if I could just remove that seconds field from the snapshot name, life would be simpler :)

BTW - you look good in that coder robe.    :)

**Comment by darkpixel** [comment]
I would love to be able to customize the snapshot names--maybe a prefix and/or suffix that gets run through through strftime.
i.e.:
```
SANOID_SNAPSHOT_PREFIX='my-sanoid-snapshots_'
SANOID_SNAPSHOT_SUFFIX='%Y-%m-%d_%H-%M-%S'
```

**Comment by filonovd** [comment]
I'd love to have the ability to customize snapshot names too. 
When working with multiple ZFS filesystems and number of snapshots having nice naming conventions that suits your tasks helps a lot. Personally I'd prefer the ability to get rid of 'autosnap_' part and have the periodic name first - like zpool/zfs@monthly-2021-03-01
It would help a lot when one wants to get all dailies for example - you just "zfs list -t snap |grep zpool/zfs@daily" 
Of course, still possible with a bit of regexping. But have it simplified would be nice.

**Comment by beren12** [comment]
yeah I would love the ability to set names as well. Sure we can can hack the code but it's a lot of effort… I'd like to be able to do prefix-daily/week/etc-time, or similar. It makes it even better with samba shadow copies, because you can use wildcards before the date but not after

**Comment by split-n** [comment]
+1 for customizable snapshot name. It's useful for multiple reasons mentioned above.

**Comment by navierb** [comment]
This feature would be very well welcomed

**Comment by kimata** [comment]
Here is an example of renaming a snapshot to make it easier to handle with Samba's vfs_shadow_copy2.

https://github.com/kimata/sanoid/commit/5d92ede545138fa1af63e87e7ccdec313f46d297

After applying the above changes, the following settings in smb.conf should work.


```
     shadow:format       = _UTC_%Y-%m-%d_%H:%M:%S                                                                                                                                                          
     shadow:snapprefix   = ^autosnap_\(yearly\)\{0,1\}\(monthly\)\{0,1\}\(weekly\)\{0,1\}\(daily\)\{0,1\}\(hourly\)\{0,1\}\(frequently\)\{0,1\}                                                            
     shadow:delimiter    = _UTC_                                                                                                                                                                           
     shadow:localtime    = no
```

**Comment by psy0rz** [comment]
For anyone needing this functionality, if you're willing to test [zfs-autobackup](https://github.com/psy0rz/zfs_autobackup), its available in v3.1.1.

**Comment by ovizii** [comment]
> For anyone needing this functionality, if you're willing to test [zfs-autobackup](https://github.com/psy0rz/zfs_autobackup), its available in v3.1.1.

I just checked and couldn't find any hint on the page you linked about snapshot naming format.

**Comment by psy0rz** [comment]
Sorry, I have yet to update the README file, i'll do that later this week.

If you have v3.1.1 you can find it in the --help
```
  --snapshot-format FORMAT
                        Snapshot naming format. Default: {}-%Y%m%d%H%M%S
  --property-format FORMAT
                        Select property naming format. Default: autobackup:{}
  --hold-format FORMAT  Hold naming format. Default: zfs_autobackup:{}

```

This internally uses pythons time.strftime() and time.strptime() to create and parse the date/time strings.

**Comment by secabeen** [comment]
```  --snapshot-format FORMAT```

Is this smart enough to handle a change in format?

**Comment by psy0rz** [comment]
Partially: It wont recognize the old snapshots as its own anymore. (so it will never delete those, you have to manually do that)

But it can continue from any common snapshot, so new stuff will be fine.

**Comment by mreymann** [comment]
Would be great if sanoid supported this. Would make it easy to purge snapshots that were not created by sanoid itself. I'll try to hack sanoid to to this. ;-)

**Comment by DaKarli** [comment]
Hi @jimsalterjrs 
as already stated by different people, it would be great if Sanoid was able to give snapshots a flexible custom name.
But even when you are not seeking for the complete solution, it would be great to change just one thing in the naming scheme or to give an option to just switch between them.

Depending on the OS you use, some symbols/characters are reserved for the specific system and cannot be used within file/folder names. For Windows one of these charecters is the colon `" : "`
Unfortunately Sanoid uses exactly this character to separate `hours : minutes : seconds` in the filename.
Due to this, Windows is not able to display these snapshots in the `.zfs/Snapshots` folder when viewing with the Windows File Explorer and they cannot be seen in the "Previous Version" function of Windows.
So unfortunately Sanoid is useless for Windows users - and this is a big user base you should not ignore.

Looking at your code I could find the related function at about line 1017 within the `"sub get_date"` block.
One could simply exchange the colons there against dots and voila, Windows can immediately see the files in the `.zfs/Snapshots` folder.
To go one step further and to enable the snapshots to be even available in the Windows Previous Files function, they must follow this naming scheme: `@GMT-2022.03.22-11.27.11` (the "@GMT is a fixed expression - from what I could find - ant the rest is clear, I suppose).

Of course, I could change these things manually in the code. But that leads to another thing I have to remember when I setup a new system or when there is an update to my base system which overrides this change - I think you know, this can be a PITA.
So for all people the best solution would be to have this change implemented at a central point and this is here.

Thanks and have a nice day! ;-)

**Comment by ams-tschoening** [comment]
I've read in some other issue being years old about individual sponsorship of features. Would that be an option in this case as well? If so, how much would it cost for which implemented feature set? For me, the following aspects would be of interest:

* custom prefix and suffix
  * e.g. replace `autosnap` with some custom wording
  * e.g. move labels like `frequently` to the front
* custom format of timestamps
  * e.g. to remove `:`
  * e.g. to make format overall shorter
* maintain formats in the config file
  * templates + overrides per dataset would be great
  * I put config files those under version control mostly

What I don't really need is when formats are changed, that old formats are taken into account or migrated or cleaned up or stuff. Should make implementation easier and I'm OK with manually cleaning up old snapshots. I'm not using syncs currently.

**Comment by rokyo249** [comment]
I also recently ran into this problem where Sanoid snapshots seem to be incompatible with TrueNAS replication, but for a different reason:

When pulling a dataset that has sub-datasets via TrueNAS replication from a server not running TrueNAS, it seems to expect the snapshots of the sub-datasets to not only conform to the naming schema (in case of Sanoid that is autosnap_%Y-%m-%d_%H:%M:%S_daily) but also have the exact same timestamp in the sub-datasets than in the main-dataset.

But when taking snapshots with Sanoid (at least on my server) it takes Sanoid a couple of seconds to take the snapshots, so the timestamps of all sub-datasets are off a few seconds from the main dataset and from each other. In my case the main dataset is called "home" and contains the /home folders of users in the naming schema "firtname.lastname" and they end up like this:

storage-fast/home@autosnap_2024-06-08_00:08:03_daily
storage-fast/home/first.user@autosnap_2024-06-08_00:08:35_daily
storage-fast/home/second.user@autosnap_2024-06-08_00:09:07_daily
storage-fast/home/third.user@autosnap_2024-06-08_00:09:51_daily
etc

Now TrueNAS refuses to pull the "home" dataset because it expects all the snapshots of its sub-datasets to be named exactly "autosnap_2024-06-08_00:08:03_daily" but they aren't.

Is there a way (or could there be) to tell Sanoid to name all snapshots of sub-datasets with the timestamp of when the snapshotting of the main dataset started instead of the timestamp from when each individual sub-dataset started?

I asked this feature in parallel on the TrueNAS forums, as I'm sure lots of people use both, TrueNAS and Sanoid, and could benefit.

This incompatibility would likely also resolve if one could omit seconds and minutes from the snapshot timestamp since (at least on my example case) the snapshotting of the entire dataset including sub-datasets never takes more than a few minutes.

**Comment by rokyo249** [comment]
Oh okay, found out that this is called "atomic" recursion and has been discussed there before (https://github.com/jimsalterjrs/sanoid/issues/244, https://github.com/jimsalterjrs/sanoid/issues/104) and also that there is a "secret" flag for doing it (first comment on: https://www.reddit.com/r/zfs/comments/tvzwc2/sanoid_needs_20_minutes_for_a_recursive_snapshot/). I'll just do that, then.

**Comment by jeffreyalles** [comment]
@sswirski Has changing the snapshot recursion to `zfs` worked consistently for you? I was running into the same issue. I changed the Sanoid configuration to `zfs`. That worked for about a week but the issue has reoccurred.

**Comment by inferenceus** [comment]
I just implemented Sanoid on my production server and immediately realised that the snapshot naming scheme is not configurable. The default naming scheme is causing difficulty with my workflow regarding ZFS maintenance.

If this is not going to be configurable, is there at least a simple way to change this in the source code? It is an essential feature for me.

**Comment by genpfault** [comment]
Timestamp construction [lives in `get_date()`](https://github.com/jimsalterjrs/sanoid/blob/v2.3.0/sanoid#L1151-L1164):

```perl
sub get_date {
	my %datestamp;
	($datestamp{'sec'},$datestamp{'min'},$datestamp{'hour'},$datestamp{'mday'},$datestamp{'mon'},$datestamp{'year'},$datestamp{'wday'},$datestamp{'yday'},$datestamp{'isdst'}) = localtime(time);
	$datestamp{'year'} += 1900;
	$datestamp{'unix_time'} = (((((((($datestamp{'year'} - 1971) * 365) + $datestamp{'yday'}) * 24) + $datestamp{'hour'}) * 60) + $datestamp{'min'}) * 60) + $datestamp{'sec'};
	$datestamp{'sec'} = sprintf ("%02u", $datestamp{'sec'});
	$datestamp{'min'} = sprintf ("%02u", $datestamp{'min'});
	$datestamp{'hour'} = sprintf ("%02u", $datestamp{'hour'});
	$datestamp{'mday'} = sprintf ("%02u", $datestamp{'mday'});
	$datestamp{'mon'} = sprintf ("%02u", ($datestamp{'mon'} + 1));
	$datestamp{'noseconds'} = "$datestamp{'year'}-$datestamp{'mon'}-$datestamp{'mday'}_$datestamp{'hour'}:$datestamp{'min'}";
	$datestamp{'sortable'} = "$datestamp{'noseconds'}:$datestamp{'sec'}";
	return %datestamp;
}
```

In particular the `noseconds` and `sortable` keys, [see the patches here](https://bezoar.org/posts/2025/1204/sanoid-time-format/).

**Comment by inferenceus** [comment]
> Timestamp construction [lives in `get_date()`](https://github.com/jimsalterjrs/sanoid/blob/v2.3.0/sanoid#L1151-L1164):
> 
> sub get_date {
> 	my %datestamp;
> 	($datestamp{'sec'},$datestamp{'min'},$datestamp{'hour'},$datestamp{'mday'},$datestamp{'mon'},$datestamp{'year'},$datestamp{'wday'},$datestamp{'yday'},$datestamp{'isdst'}) = localtime(time);
> 	$datestamp{'year'} += 1900;
> 	$datestamp{'unix_time'} = (((((((($datestamp{'year'} - 1971) * 365) + $datestamp{'yday'}) * 24) + $datestamp{'hour'}) * 60) + $datestamp{'min'}) * 60) + $datestamp{'sec'};
> 	$datestamp{'sec'} = sprintf ("%02u", $datestamp{'sec'});
> 	$datestamp{'min'} = sprintf ("%02u", $datestamp{'min'});
> 	$datestamp{'hour'} = sprintf ("%02u", $datestamp{'hour'});
> 	$datestamp{'mday'} = sprintf ("%02u", $datestamp{'mday'});
> 	$datestamp{'mon'} = sprintf ("%02u", ($datestamp{'mon'} + 1));
> 	$datestamp{'noseconds'} = "$datestamp{'year'}-$datestamp{'mon'}-$datestamp{'mday'}_$datestamp{'hour'}:$datestamp{'min'}";
> 	$datestamp{'sortable'} = "$datestamp{'noseconds'}:$datestamp{'sec'}";
> 	return %datestamp;
> }
> 
> In particular the `noseconds` and `sortable` keys, [see the patches here](https://bezoar.org/posts/2025/1204/sanoid-time-format/).

@genpfault In particular, it's more the `autosnap_` and `_<frequency>` components which are problematic, along with colons rather than continued hyphenation for times after dates. I would rather have only the date and time, as I did via cron before migrating to Sanoid. Is it as simple as just finding and changing those naming schemes in the code?

**Comment by araemo** [comment]
I'd like to add my request to allow snapshot naming without colons.

My use case:
I am trying to leverage my existing ZFS file/VM server to back up proxmox VMS that are on ZFS datastores on the proxmox hosts.

Proxmox's built-in snapshot feature leverages ZFS snapshots, but it does not support naming them with colons.  I'd REALLY like to be able to simply use syncoid to replicate them to another host which uses sanoid to prune old snapshots, but keep more on the remote host than are on the local host.  (It seems to me that leveraging zfs snapshots/zfs-replication natively will be better for my homelab use case than proxmox backup server as a VM on top of my ZFS pool)

If I could configure sanoid to use something other than colons for the timestamps, I could create snapshots within proxmox (so, supporting things like flushing the disks/pausing the VM before taking the snapshot), but have sanoid take care of removing old snapshots from the backup target, rather than only keeping the exact same snapshots on the VM host and the remote backup host.

Of course, I also already have many datasets using the existing naming convention for their snapshots, so being able to override this per template/dataset would be ideal.
